wordpress-mobile/WordPress-iOS
# Add publish functionality to Gutenberg editor

### Description

This change adds Publish functionality to the Gutenberg editor. Related issue: https://github.com/wordpress-mobile/gutenberg-mobile/issues/182

![pr_demo](https://user-images.githubusercontent.com/5032900/48556088-faf72d00-e8f3-11e8-88c1-b7d4cbc785eb.gif)

Below buttons gain new functionality with this change:

![simulator screen shot - iphone 7 - 2018-11-15 at 16 09 13](https://user-images.githubusercontent.com/5032900/48555047-20366c00-e8f1-11e8-9255-2d741110a4ef.png)

### Implementation Details

The basic idea of the implementation is to move common functionality to PostEditorUtil and use it from both GutenbergViewController and AztecPostViewController. So the change set mostly consists of cut/paste from AztecPostViewController to PostEditorUtil. Despite of being big it actually doesn't change a lot since all its doing is moving the logic to another place and solving the compile errors resulting from that. I tested AztecPostViewController by comparing the screen on 2 simulators (one is from master branch the other one is from current branch) and it looks OK but I highly encourage reviewers to do the same king of comparison test.

On the other hand, if this looks risky to you we can just reset everything we did to AztecPostViewController, the Gutenberg side wouldn't be affected from that but it'll cause duplicate logic so I would go with common utility where possible.


### Test

**Testing prerequisites**:

- Checkout the PRs branch
- clean pods
- clean build folder on XCode
- run pod install
- cd to WordPress-iOS/Gutenberg, run yarn install, yarn start:reset
- Run WordPress-iOS app via XCode 

I also **strongly** encourage opening another iOS simulator that runs the code from the master branch. It gives opportunity to compare each part and saving lots of time when we need to decide if sth is the default behavior.
 
**Keep in mind**:

1. currently Publish(Or Update) button is always active on Gutenberg editor unlike Aztec editor. This is done intentionally because Gutenberg side doesn't have the capability of informing every single change yet.

2. If the post is new or Draft you'll see button label as "Publish", but if you try to edit a published post you 'll see it as "Update". This can be confusing but it is actually a design intent.

**Gutenberg tests**

Test 1 - Edit published post:

- Open one of the published posts
- Edit something in the post
- Tap Update button
- Verify that a loading animation pops
- Tap close button
- Reopen the post and verify that it is changed

Test 2: Close button

Test 2.1 - Close edited post(which was published before) with "Publish" option:

- Open one of the published posts
- Edit something in the post
- Tap close button
- Verify that you see the "You have unsaved changes" alert sheet
- Tap Update post
- Verify that you see the "Are you sure you want to publish?" alert sheet
- Tap "Publish"
- Verify that editor is closed and you see an updating animation below the post summary

Test 2.2 - Attempt Close edited post(which was published before) and then "Keep editing":

- Open one of the published posts
- Edit something in the post
- Tap close button
- Verify that you see the "You have unsaved changes" alert sheet
- Tap "Keep editing"
- Verify that alert sheet closes and the content doesn't change

Test 2.3 - Close edited post(which was published before) with "Discard" option:

- Open one of the published posts
- Edit something in the post
- Tap close button
- Verify that you see the "You have unsaved changes" alert sheet
- Tap "Discard"
- Verify that editor is closed and your last changes aren't saved.

Test 2.4 Close un-edited post

- Open one of the published posts
- Tap close button
- Verify that editor is closed without a prompt and nothing changes in the post

**Aztec tests**

Open EditorSettings.swift in branch try/gutenberg-publish.

Update this line    

  `private var current: Editor {
        return .gutenberg
    }`

as 
    `private var current: Editor {
        return .aztec
    }`

Reopen the WordPress-iOS app in try/gutenberg-publish.

Test and compare below cases between WordPress-iOS master and try/gutenberg-publish.

- Create a new post and Save as draft
- Create a new post and Publish

- Edit a draft post and Update
- Edit a draft post, tap Close button and Discard
- Edit a draft post, tap Close button and Update Draft
- Edit a draft post, tap Close button and Keep Editing
- Edit a draft post and select navigation bar right button (...)->"Publish Now"

- Edit a published post and Update
- Edit a published post, tap Close button and Discard
- Edit a published post, tap Close button and Update Draft
- Edit a published post, tap Close button and Keep Editing

- The change doesn't affect this directly but trying also uploading a media, adding a link and other things and publishing would be good.
